kotlinx-coroutines-core/common/src/flow/operators/Context.kt

```diff
@@ -109,10 +109,15 @@ public fun <T> Flow<T>.buffer(capacity: Int = BUFFERED): Flow<T> {
     require(capacity >= 0 || capacity == BUFFERED || capacity == CONFLATED) {
         "Buffer size should be non-negative, BUFFERED, or CONFLATED, but was $capacity"
     }
-    return if (this is ChannelFlow)
-        update(capacity = capacity)
-    else
-        ChannelFlowOperatorImpl(this, capacity = capacity)
+
+    return when (this) {
+        is ChannelFlowOperatorImpl -> update(
+            capacity = capacity,
+            dropWhileBusy = if (capacity == CONFLATED) false else dropWhileBusy
```

Comment:
I would rather move the logic of merging `dropWhileBusy` flag into the implementation of `update` function and leave the code in the `buffer` operator as it was before. It will just call `update(capacity = capacity)` leaving a default value of `dropWhileBusy = false` and letting `update` take care of figuring it out.